packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,507 @@
+  List<CurvedAnimation> _endAnimations;
+  List<CurvedAnimation> _gapAnimations;
+
+  void _initSlice(int index) {
+    final AnimationController startController = new AnimationController(
+      duration: kThemeAnimationDuration
+    );
+    final AnimationController endController = new AnimationController(
+      duration: kThemeAnimationDuration
+    );
+
+    _startControllers[index] = startController;
+    _endControllers[index] = endController;
+
+    final CurvedAnimation startAnimation = new CurvedAnimation(
+      parent: startController,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+    final CurvedAnimation endAnimation = new CurvedAnimation(
+      parent: endController,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+
+    _startAnimations[index] = startAnimation;
+    _endAnimations[index] = endAnimation;
+
+    startAnimation.addListener(_handleTick);
+    endAnimation.addListener(_handleTick);
+  }
+
+  void _initGap(int index) {
+    final AnimationController gapController = new AnimationController(
+      duration: kThemeAnimationDuration
+    );
+
+    _gapControllers[index] = gapController;
+
+    final CurvedAnimation gapAnimation = new CurvedAnimation(
+      parent: gapController,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+
+    _gapAnimations[index] = gapAnimation;
+
+    gapAnimation.addListener(_handleTick);
+  }
+
+  void _insertNulls(int index) {
+    _startControllers.insert(index, null);
+    _endControllers.insert(index, null);
+    _gapControllers.insert(index, null);
+    _startAnimations.insert(index, null);
+    _endAnimations.insert(index, null);
+    _gapAnimations.insert(index, null);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _startControllers = new List<AnimationController>();
+    _endControllers = new List<AnimationController>();
+    _gapControllers = new List<AnimationController>();
+    _startAnimations = new List<CurvedAnimation>();
+    _endAnimations = new List<CurvedAnimation>();
+    _gapAnimations = new List<CurvedAnimation>();
+
+    for (int i = 0; i < children.length; i += 1) {
+      _insertNulls(i);
+
+      if (children[i].isGap) {
+        _initGap(i);
+        _gapControllers[i].value = 1.0; // Gaps must be wide when first built.
+      } else {
+        _initSlice(i);
+      }
+    }
+    _assertGaps(children);
+    _fixSliceStates(children);
+  }
+
+  @override
+  void dispose() {
+    for (AnimationController controller in _startControllers)
+      controller?.dispose();
+    for (AnimationController controller in _endControllers)
+      controller?.dispose();
+    for (AnimationController controller in _gapControllers)
+      controller?.dispose();
+    super.dispose();
+  }
+
+  void _handleTick() {
+    setState(() {
+      // The animation's state is our build state, and it changed already.
+    });
+  }
+
+  bool _consecutiveGaps(List<MergeableMaterialItem> children) {
```

Comment:
this is an expensive function. Put the assert in this function, and call it debugAssertNoConsecutiveGaps or something.